src/main/kotlin/com/makarytskyi/rentcar/service/impl/OrderServiceImpl.kt

```diff
@@ -87,23 +92,17 @@ internal class OrderServiceImpl(
     private fun validateUserExists(userId: String) = userRepository.findById(userId)
         .switchIfEmpty { Mono.error(IllegalArgumentException("User with id $userId is not found")) }
 
-    private fun validateCarAvailability(carId: String?, from: Date?, to: Date?): Mono<MongoOrder> {
-        return Mono.justOrEmpty(carId)
-            .switchIfEmpty { Mono.error(IllegalArgumentException("Car id should not be null")) }
-            .flatMap { carRepository.findById(it) }
+    private fun validateCarAvailability(carId: String, from: Date?, to: Date?): Mono<MongoOrder> {
+        return carRepository.findById(carId)
             .switchIfEmpty { Mono.error(NotFoundException("Car with id $carId is not found")) }
             .flatMapMany { orderRepository.findByCarId(it.id.toString()) }
-            .handle<MongoOrder> { order, sink ->
-                if (order.from?.before(to) == true && order.to?.after(from) == true) {
-                    sink.error(IllegalArgumentException("Order on these dates is already exist"))
-                }
-            }
-            .next()
+            .filter { it.from?.before(to) == true && it.to?.after(from) == true }
+            .flatMap<MongoOrder> { Mono.error(IllegalArgumentException("Car is already booked on this time")) }
+            .toMono()
```

Comment:
why toMono if you're already have mono chain?

Reply:
Because, I used flatMapMany before and it returns Flux without toMono.